modules/2_mnist/src/mnist.cpp

```diff
@@ -1,70 +1,90 @@
-    // at http://yann.lecun.com/exdb/mnist/
+    labels = lblDispatcher.getAllLabels();
 }
 
 void prepareSamples(const std::vector<cv::Mat>& images, cv::Mat& samples) {
-    CV_Error(Error::StsNotImplemented, "prepareSamples");
+    // assume each image has the same constraints
+    cv::Size imgConstraints = images[0].size();
+    samples = cv::Mat(images.size(), imgConstraints.width * imgConstraints.height, CV_8U);
+
+    for (int i = 0; i < images.size(); i++) {
+        images[i].reshape(1, 1).copyTo(samples.row(i));
+    }
+
+    samples.convertTo(samples, CV_32F);
 }
 
-Ptr<ml::KNearest> train(const std::vector<cv::Mat>& images,
-                        const std::vector<int>& labels) {
-    CV_Error(Error::StsNotImplemented, "train");
+cv::Ptr<cv::ml::KNearest> train(const std::vector<cv::Mat>& images,
+        const std::vector<int>& labels) {
+    cv::Ptr<cv::ml::KNearest> knn(cv::ml::KNearest::create());
+
+    cv::Mat samples;
+    prepareSamples(images, samples);
+
+    knn->train(samples, cv::ml::SampleTypes::ROW_SAMPLE, labels);
+
+    return knn;
 }
 
-float validate(Ptr<ml::KNearest> model,
+float validate(const cv::Ptr<cv::ml::KNearest>& model,
                const std::vector<cv::Mat>& images,
                const std::vector<int>& labels) {
-    CV_Error(Error::StsNotImplemented, "validate");
+    int correctGuesses = 0;
+
+    cv::Mat samples;
+    prepareSamples(images, samples);
+
+    cv::Mat resultMat;
+    model->findNearest(samples, model->getDefaultK(), resultMat);
+
+    for (int i = 0; i < images.size(); i++) {
+        correctGuesses += resultMat.at<int>(0, i) == labels[i];
+    }
+
+    return (float)correctGuesses / images.size() * 100;
 }
 
-int predict(Ptr<ml::KNearest> model, const Mat& image) {
-    // TODO: resize image to 28x28 (cv::resize)
+int predict(const cv::Ptr<cv::ml::KNearest>& model, const cv::Mat& image) {
+    cv::Mat image_hsv;
+    // resize
+    cv::resize(image, image_hsv, cv::Size(28, 28));
 
-    // TODO: convert image from BGR to HSV (cv::cvtColor)
+    // bgr to hsv
+    cv::cvtColor(image_hsv, image_hsv, cv::COLOR_BGR2HSV);
 
-    // TODO: get Saturate component (cv::split)
+    // split hsv, hsvComponents[1] holds the saturation component
+    std::vector<cv::Mat> hsvComponents(3);
+    cv::split(image_hsv, hsvComponents);
 
-    // TODO: prepare input - single row FP32 Mat
+    // flatten & convert to unsigned byte
+    cv::Size imageSize = hsvComponents[1].size();
+    cv::Mat preparedImage(1, imageSize.width * imageSize.height, CV_8U);
 
-    // TODO: make a prediction by the model
+    hsvComponents[1].reshape(1, 1).copyTo(preparedImage.row(0));
 
-    CV_Error(Error::StsNotImplemented, "predict");
-}
+    preparedImage.convertTo(preparedImage, CV_32F);
+
+    // make a prediction by the model
+    cv::Mat resultMat;
+    model->findNearest(preparedImage, model->getDefaultK(), resultMat);
+
+    return resultMat.at<int>(0, 0);
```

Comment:
According to CI build (https://travis-ci.com/github/dkurt/openvino_practice/builds/175474986):
```
Predicted numbers: 1086324736 1086324736 1088421888
/home/travis/build/dkurt/openvino_practice/modules/2_mnist/test/test_model.cpp:47: Failure
Value of: ref[i]
  Actual: 6
Expected: out[i]
Which is: 1086324736
```
it seems to me that `resultMat` is not `CV_32SC1` and we should not get element as `int`. Probably it's `CV_32FC` and you need to get as `resultMat.at<float>(0, 0);`

Reply:
Thanks a lot, @dkurt , it works now!